### DISCUSSION TITLE
Is it possible to use getStaticProps() and serve two different versions of the same page based on the domain/headers?

### DISCUSSION THREAD

**Discussion by ZinoKader** [vercel/next.js#21546] [Help]
An example of a usecase for this would be to have a very heavy and slowly loading page that needs to be statically built, so getStaticProps is a must - but this page also requires some kind of authentication. Suppose this page should then be able to show two different versions of the same page to unauthenticated/authenticated users. Also suppose the authentication is determined by which domain or which headers the user accesses the page with. Lastly suppose we don't want to use a different path for the two versions of the page.

I don't know if this is supported in any way by Next.js right now but it seems like it should be possible by building two versions of a page at build time and letting the Next server map the same path/URL to a different pre-built version of the page based on the authentication described above.

Is this possible right now?

**Comment by neighlyd-alo** [comment]
I am trying to find a solution to this issue as well. Imagine a page that is a marketing page when signed out, but is a contentful page when signed in. Sending through a skeleton to both and populating defeats the benefits of SSG entirely (both in terms of SEO and TTI). Conversely, sending through the pre-rendered marketing page to signed in users is a horrible experience. 

Currently, I have four possible "solutions." However, I don't really like any of them:

1) SSR the page, giving the renderer access to auth state, and then have CDN heavily cache the page if no auth cookie is present. This solution has the upside of giving us two different pages, but the downside is that it relies on the CDN making the right choice and also risks heavy server spikes on signed-in SSR.
2) SSG the signed out page and hide it behind some css that triggers around a client-side `getUser()` action. This will reduce the network load on client-side needed to expose the content, however it will cause signed-in users to get a substantial page that they won't even use and _then_ will have to download the data they need for their content.
3) Use something like NGINX to remap requests based on cookie presence. So, signed-out requests get passed through normally, signed-in requests get remapped to `domain.com/app/${route}`. Like number 1, this adds a layer of indirection, and it also increases the complexity of the `/pages` directory routing scheme. I am also not entirely sure how well it will work with things like `<Link />` and various preloading optimizations. This method should be opaque to users (i.e. URL path does not change). 
4) Similar to 3, we could use our proxy server to send redirects back to users who are signed-in (based on auth cookie). This will place the user in `/app/` land, with all the up and downsides listed in #3. However, because it's an explicit redirect, it means that it will be transparent to the users.

Having a method to handle these states built in to Next would be preferable, but I don't see any way to accomplish these tasks with its current design. If anyone has any ideas, I would love to hear them.

**Reply by sorokinvj** [reply]
I do have a similar problem, which I can abstract as "How to render UI user preferences with SSG". 

Imagine you have a simple blog with two views - card and list, and we keep the user preference in cookies or in local storage. All the data we need we do have at a build time, so SSG is our choice. However, we render only one view (suppose card) with SSG, so for users that choose another view (i.e list), there would be a flash of content. 

Thinking of solving this blows my mind – I just can't understand how do you wait for user preferences and not show default statically generated layout. Keeping two versions at once seems excessive, but imagine your app is bigger and has a lot of UI options (dark theme, data filters etc), how do you approach it in general?

**Comment by fabb** [comment]
We do this in our custom express server with a custom middleware. We need it for showing pages without header/footer when they are displayed in a webview of our native app. We basically just create 2 SSG pages, while the second one is located in a folder `pages//internal/` and cannot be directly opened from the browser (we have another express middleware blocking these requests). In our middleware, we check headers, and when we detect a relevant header, we modify the url of the request and call the render method on the next server instance. If there is no such header, we just ignore the request and let it be handled regularly.
Note that there is an issue with this approach: client routing. You need to rewrite client routed urls too, or make sure that you only use server routing in case the header is present. Otherwise you will still be able to navigate to pages where it shouldn’t be possible.
